src/profile-logic/profile-data.js

```diff
@@ -569,6 +569,234 @@ export function filterThreadToPrefixCallNodePath(
+    const stacksChecked: Array<void | true> = [];
+
+    // Go through each sample and determine if it contains a stack that needs to be
+    // merged.
+    for (let i = 0; i < samples.stack.length; i++) {
+      const leafStackIndex = samples.stack[i];
+      if (leafStackIndex === null || stacksChecked[leafStackIndex]) {
+        continue;
+      }
+      stacksChecked[leafStackIndex] = true;
+
+      let matchesUpToDepth = 0; // counted from the leaf
+      for (
+        let stackIndex = leafStackIndex;
+        stackIndex !== null;
+        stackIndex = stackTable.prefix[stackIndex]
+      ) {
+        const frameIndex = stackTable.frame[stackIndex];
+        const funcIndex = frameTable.func[frameIndex];
+
+        if (funcIndex === postfixCallNodePath[matchesUpToDepth]) {
+          // The CallNodePath matches up to this depth.
+          matchesUpToDepth++;
+          if (matchesUpToDepth === postfixDepth) {
+            // This matches the CallNodePath.
+            stackNeedsMerging[stackIndex] = true;
+            break;
+          }
+        } else if (funcMatchesImplementation(thread, funcIndex)) {
+          // This function didn't match the CallNodePath, and it can't be ignored
+          // because it matches the implementation.
+          break;
+        }
+      }
+    }
+
+    const oldStackToNewStack: Map<
+      IndexIntoStackTable | null,
+      IndexIntoStackTable | null
+    > = new Map();
+    oldStackToNewStack.set(null, null);
+    const newStackTable = {
+      length: 0,
+      prefix: [],
+      frame: [],
+    };
+
+    // We have determined which stacks need to be merged, now do the merging in
+    // one pass across all the stacks.
+    for (
+      let oldStackIndex = 0;
+      oldStackIndex < stackTable.length;
+      oldStackIndex++
+    ) {
+      const oldPrefix = stackTable.prefix[oldStackIndex];
+      const newPrefix = oldStackToNewStack.get(oldPrefix);
+      if (newPrefix === undefined) {
+        throw new Error('The stack must not have an undefined prefix.');
+      }
+      // Skip over this stack, since we are merging it.
+      if (stackNeedsMerging[oldStackIndex]) {
+        oldStackToNewStack.set(oldStackIndex, newPrefix);
+        continue;
+      }
+
+      const newStackIndex = newStackTable.length++;
+      newStackTable.prefix.push(newPrefix);
+      newStackTable.frame.push(stackTable.frame[oldStackIndex]);
+      oldStackToNewStack.set(oldStackIndex, newStackIndex);
+    }
+
+    const newSamplesTable = Object.assign({}, samples, {
+      stack: samples.stack.map(oldStack => {
+        const newStack = oldStackToNewStack.get(oldStack);
+        if (newStack === undefined) {
+          throw new Error('The stack must not convert to undefined.');
+        }
+        return newStack;
+      }),
+    });
+
+    return Object.assign({}, thread, {
+      stackTable: newStackTable,
+      samples: newSamplesTable,
+    });
+  });
+}
+
+const FUNC_MATCHES = {
```

Comment:
I need to follow up and use this anywhere else where we are checking the implementation. We get bad results on C++ only implementations.